File: library/c/quadraturerules.c
#include "quadraturerules.h"
{{for Q in rules}}
#include "{{Q.snake_case_name}}.c"
{{end for}}

int single_integral_quadrature(
    QuadratureRule rtype,
    Domain domain,
    int order,
    double* points,
    double* weights
)
{
  switch (rtype)
  {
  {{for Q in rules}}
  {{if Q.itype == single}}
  case QR_{{Q.PascalCaseName}}:
    {{Q.snake_case_name}}(domain, order, points, weights);
    return 0;
  {{end if}}
  {{end for}}
  default:
    return 1;
  }
}

int double_integral_quadrature(
    QuadratureRule rtype,
    Domain domain,
    int order,
    double* first_points,
    double* second_points,
    double* weights
)
{
  switch (rtype)
  {
  {{for Q in rules}}
  {{if Q.itype == double}}
  case QR_{{Q.PascalCaseName}}:
    {{Q.snake_case_name}}(domain, order, first_points, second_points, weights);
    return 0;
  {{end if}}
  {{end for}}
  default:
    return 1;
  }
}

int single_integral_quadrature_points_size(
    QuadratureRule rtype,
    Domain domain,
    int order
)
{
  switch (rtype)
  {
  {{for Q in rules}}
  {{if Q.itype == single}}
  case QR_{{Q.PascalCaseName}}:
    return {{Q.snake_case_name}}_points_size(domain, order);
  {{end if}}
  {{end for}}
  default:
    return -1;
  }
}

int single_integral_quadrature_weights_size(
    QuadratureRule rtype,
    Domain domain,
    int order
)
{
  switch (rtype)
  {
  {{for Q in rules}}
  case QR_{{Q.PascalCaseName}}:
    return {{Q.snake_case_name}}_weights_size(domain, order);
  {{end for}}
  default:
    return -1;
  }
}

int double_integral_quadrature_first_points_size(
    QuadratureRule rtype,
    Domain domain,
    int order
)
{
  switch (rtype)
  {
  {{for Q in rules}}
  {{if Q.itype == double}}
  case QR_{{Q.PascalCaseName}}:
    return {{Q.snake_case_name}}_first_points_size(domain, order);
  {{end if}}
  {{end for}}
  default:
    return -1;
  }
}

int double_integral_quadrature_second_points_size(
    QuadratureRule rtype,
    Domain domain,
    int order
)
{
  switch (rtype)
  {
  {{for Q in rules}}
  {{if Q.itype == double}}
  case QR_{{Q.PascalCaseName}}:
    return {{Q.snake_case_name}}_second_points_size(domain, order);
  {{end if}}
  {{end for}}
  default:
    return -1;
  }
}

int double_integral_quadrature_weights_size(
    QuadratureRule rtype,
    Domain domain,
    int order
)
{
  switch (rtype)
  {
  {{for Q in rules}}
  case QR_{{Q.PascalCaseName}}:
    return {{Q.snake_case_name}}_weights_size(domain, order);
  {{end for}}
  default:
    return -1;
  }
}

File: library/fortran90/quadraturerules.f90
{{for Q in rules}}
include "{{Q.snake_case_name}}.f90"
{{end for}}

module quadraturerules
  implicit none

  public single_integral_quadrature
  {{for D in domains}}
  public QR_{{D.PascalCaseName}}
  {{end for}}
  {{for R in rules}}
  public QR_{{R.PascalCaseName}}
  {{end for}}
  enum, bind(C)
    {{for D in domains}}
    enumerator :: QR_{{D.PascalCaseName}} = {{D.index}}
    {{end for}}
  end enum
  enum, bind(C)
    {{for R in rules}}
    enumerator :: QR_{{R.PascalCaseName}} = {{R.index}}
    {{end for}}
  end enum

contains

  subroutine single_integral_quadrature(rtype, domain, order, points, weights)
    {{for Q in rules}}
    use {{Q.snake_case_name}}
    {{end for}}
    implicit none
    integer, value :: rtype
    integer, value :: domain
    integer, value :: order
    real, allocatable, intent(out) :: points(:,:)    
    real, allocatable, intent(out) :: weights(:)

    {{for Q in rules}}
    {{for D in domains}}
    {{if Q.itype == single}}
    if (rtype == QR_{{Q.PascalCaseName}} .and. domain == QR_{{D.PascalCaseName}}) then
      call {{Q.abbrv_name}}_{{D.abbrv_name}}(order, points, weights)
    end if
    {{end if}}
    {{end for}}
    {{end for}}
  end subroutine

end module quadraturerules
